nexus/src/sagas.rs

```diff
@@ -223,3 +246,190 @@ async fn sic_instance_ensure(
+        // to undo at this point?
+        new_action_noop_undo(sim_cleanup_source),
+    );
+
+    template_builder.build()
+}
+
+async fn sim_alloc_server(
+    sagactx: ActionContext<SagaInstanceMigrate>,
+) -> Result<Uuid, ActionError> {
+    use omicron_common::api::external::{
+        ByteCount, IdentityMetadataCreateParams, InstanceCpuCount,
+    };
+
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    let migrate_type = sagactx.lookup::<MigrateType>("migrate_type")?;
+
+    if let MigrateType::InPlace = migrate_type {
+        Ok(params.migrate_params.dst_sled_uuid)
+    } else {
+        // TODO: These parameters are unused, `alloc_server` just
+        // needs to be passed something. See comment on
+        // SagaContext::alloc_server()
+        let create_params = params::InstanceCreate {
+            identity: IdentityMetadataCreateParams {
+                name: "unused".parse().unwrap(),
+                description: "unused".to_string(),
+            },
+            ncpus: InstanceCpuCount(0),
+            memory: ByteCount::from_mebibytes_u32(0),
+            hostname: "unused".to_string(),
+        };
+        osagactx
+            .alloc_server(&create_params)
+            .await
+            .map_err(ActionError::action_failed)
+    }
+}
+
+async fn sim_instance_migrate(
+    sagactx: ActionContext<SagaInstanceMigrate>,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+
+    let dst_sled_uuid = sagactx.lookup("dst_server_id")?;
+    let dst_propolis_uuid = sagactx.lookup::<Uuid>("dst_propolis_id")?;
+
+    let instance = osagactx
+        .datastore()
+        .instance_fetch_by_name(
+            &params.project_id,
+            &params.instance_name.clone().into(),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let instance_id = instance.id();
+
+    let old_runtime: InstanceRuntimeState = instance.runtime_state.into();
+    let runtime = InstanceRuntimeState {
+        sled_uuid: dst_sled_uuid,
+        propolis_uuid: dst_propolis_uuid,
+        propolis_addr: None,
+        ..old_runtime
+    };
+    let instance_hardware = sled_agent_client::types::InstanceHardware {
+        runtime: runtime.into(),
+        // TODO: populate NICs
+        nics: vec![],
+    };
+    let target = sled_agent_client::types::InstanceRuntimeStateRequested {
+        // TODO: is this the appropriate state?
+        run_state: sled_agent_client::types::InstanceStateRequested::Running,
+    };
+
+    let src_propolis_uuid = old_runtime.propolis_uuid;
+    let src_propolis_addr = old_runtime.propolis_addr.ok_or_else(|| {
+        ActionError::action_failed("expected source propolis-addr".to_string())
+    })?;
+
+    let dst_sa = osagactx
+        .sled_client(&dst_sled_uuid)
+        .await
+        .map_err(ActionError::action_failed)?;
```

Comment:
To be clear, the order of operations in this saga step is:

1. Fetch the instance from DB
2. Send "ensure + migrate" request to the destination instance, using the metadata for the instance we got in (1).
3. Update the "runtime state" in the DB

Is it possible we run into a TOCTTOU issue here?

What if someone modifies the sled (perhaps sends a different `ensure` request, gets a response) in between our "fetch" and "ensure" requests? Wouldn't we send an out-of-date sled request?

I think this problem could be avoided with some combination of:
1. Durable state changes (explicitly modifying the DB, indicating that the instance is in the progress of migration, and preventing concurrent modification until this flag is removed).
2. Using a generation number to only perform an operation if we explicitly ensure we haven't raced with anyone else.

https://rfd.shared.oxide.computer/rfd/0192#_conditional_updates seems related

Reply:
Yup, we ask sled agent to idempotently update the state to migrating (via Nexus) as part of an earlier action. Only if that action succeeds will we move on to the subsequent actions which does the actual migration.